In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
ROOT_DIR = '/content/drive/My Drive/Recommendation Systems/movielens/'
DATA_DIR = '/content/drive/My Drive/Recommendation Systems/movielens/data/'

In [0]:
import pandas as pd
import numpy as np

In [0]:
#Load the u.user file into dataframe
u_cols = ['user_id', 'age', 'sex', 'occupation', 'zip_code']

users = pd.read_csv(DATA_DIR+'u.user',sep='|',names=u_cols,encoding='latin-1')

In [0]:
users.head()

user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

In [0]:
#Load the u.items file into a dataframe
i_cols = ['movie_id', 'title' ,'release date','video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure',
'Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama','Fantasy','Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi',
'Thriller', 'War', 'Western']

movies = pd.read_csv(DATA_DIR+'u.item', sep='|', names=i_cols,encoding='latin-1')

In [0]:
movies.head()

movie_id              title release date  ...  Thriller War  Western
0         1   Toy Story (1995)  01-Jan-1995  ...         0   0        0
1         2   GoldenEye (1995)  01-Jan-1995  ...         1   0        0
2         3  Four Rooms (1995)  01-Jan-1995  ...         1   0        0
3         4  Get Shorty (1995)  01-Jan-1995  ...         0   0        0
4         5     Copycat (1995)  01-Jan-1995  ...         1   0        0

[5 rows x 24 columns]

In [0]:
#Remove all information except Movie ID and title
movies = movies[['movie_id', 'title']]

In [0]:
#Load the u.data file into a dataframe
r_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_csv(DATA_DIR+'u.data', sep='\t', names=r_cols,encoding='latin-1')
ratings.head()

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [0]:
#Drop the timestamp column
ratings = ratings.drop('timestamp', axis=1)

In [0]:
from sklearn.model_selection import train_test_split

X = ratings.copy()
y= ratings['user_id']

#Split into training and test datasets, stratified along user_id
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25,stratify=y, random_state=42)

In [0]:
ratings['user_id'].nunique()

943

In [0]:
print(X_train['user_id'].nunique())
print(X_test['user_id'].nunique())

943
943


In [0]:
#Import the mean_squared_error function
from sklearn.metrics import mean_squared_error
#Function that computes the root mean squared error (or RMSE)
def rmse(y_true, y_pred):
  return np.sqrt(mean_squared_error(y_true, y_pred))

In [0]:
def baseline(user_id,movie_id):
  return 3.0

In [0]:
def score(cf_model):
  id_pairs = zip(X_test['user_id'],X_test['movie_id'])
  y_pred = np.array([cf_model(user,movie) for (user,movie) in id_pairs])
  y_true = np.array(X_test['rating'])
  return rmse(y_true,y_pred)

In [0]:
score(baseline)

1.2470926188539486

For any models that we'll build, the rmse should be less than the baseline's rmse 1.247

In [0]:
import pickle

with open(DATA_DIR+'X_train.pkl','wb') as f:
  pickle.dump(X_train,f)
with open(DATA_DIR+'X_test.pkl','wb') as f:
  pickle.dump(X_test,f)
with open(DATA_DIR+'y_train.pkl','wb') as f:
  pickle.dump(y_train,f)
with open(DATA_DIR+'y_test.pkl','wb') as f:
  pickle.dump(y_test,f)